src/main/host/memory_manager/memory_mapper.rs

```diff
@@ -0,0 +1,1032 @@
+                }
+                Mutation::Removed(modified_interval, mut modified_region) => {
+                    modified_region.prot = prot;
+                    if !modified_region.shadow_base.is_null() {
+                        unsafe {
+                            sys::mman::mprotect(
+                                modified_region.shadow_base,
+                                modified_interval.len(),
+                                protflags,
+                            )
+                        }
+                        .unwrap_or_else(|e| warn!("mprotect: {}", e));
+                    }
+                    assert!(self
+                        .regions
+                        .insert(modified_interval, modified_region)
+                        .is_empty());
+                }
+            }
+        }
+        Ok(0.into())
+    }
+
+    // Get a raw pointer to the plugin's memory, if it's been remapped into Shadow.
+    // Panics if called with zero-length `src`.
+    fn get_mapped_ptr<T: Pod + Debug>(&self, src: TypedPluginPtr<T>) -> Option<*mut T> {
+        assert!(src.len() > 0);
+
+        let (interval, region) = match self.regions.get(usize::from(src.ptr())) {
+            Some((i, r)) => (i, r),
+            None => {
+                warn!("src {:?} isn't in any mapped region", src);
+                return None;
+            }
+        };
+        let shadow_base = if region.shadow_base.is_null() {
+            trace!("src {:?} isn't mapped into Shadow", src);
+            return None;
+        } else {
+            region.shadow_base
+        };
+
+        if !interval.contains(&(usize::from(src.slice(src.len()..src.len()).ptr()) - 1)) {
+            // End isn't in the region.
+            trace!(
+                "src {:?} mapped into Shadow, but extends beyond mapped region.",
+                src
+            );
+            return None;
+        }
+
+        let offset = usize::from(src.ptr()) - interval.start;
+        // Base pointer + offset won't wrap around, by construction.
+        let ptr = unsafe { shadow_base.add(offset) } as *mut T;
+
+        Some(ptr)
+    }
+
+    fn get_mapped_ptr_and_count<T: Pod + Debug>(&self, src: TypedPluginPtr<T>) -> Option<*mut T> {
+        let res = self.get_mapped_ptr(src);
+        if res.is_none() {
+            self.inc_misses(src);
+        }
+        res
+    }
+
+    pub unsafe fn get_ref<T: Debug + Pod>(&self, src: TypedPluginPtr<T>) -> Option<&[T]> {
+        if src.len() == 0 {
+            return Some(&[]);
+        }
+        let ptr = self.get_mapped_ptr_and_count(src)?;
+        Some(unsafe { std::slice::from_raw_parts(ptr, src.len()) })
+    }
+
+    pub unsafe fn get_mut<T: Debug + Pod>(&self, src: TypedPluginPtr<T>) -> Option<&mut [T]> {
+        if src.len() == 0 {
+            return Some(&mut []);
+        }
+        let ptr = self.get_mapped_ptr_and_count(src)?;
+        Some(unsafe { std::slice::from_raw_parts_mut(ptr, src.len()) })
```

Comment:
Added some standalone helper functions that simply check and return the pointer - basically just letting us inline the assertions at point of use instead of having to use a separate statement.